### DISCUSSION TITLE
Binding for ScrollbarChart?

### DISCUSSION THREAD

**Discussion by houstonhaynes** [JordanMarr/ReactiveElmish.Avalonia#45] [Q&A]
I'm essentially trying to build something like [this](https://livecharts.dev/docs/Avalonia/2.0.0-rc2/samples.general.scrollable). 

It's been a bit touchy getting things to show up as expected. It's a real stretch for me but I figured it would be a way to get my head wrapped around MVVM to MVU "logic" as well as bang my head against Avalonia a bit. I hacked my way into the current state and then started asking GitHub Copilot for help, with unfortunately expected results. Right now it seems that there's an issue where an event is raised before the View is fully initialized. As I "leave" and return to the view then the MVU loop seems to trigger as expected (using the trusty old console log method). I figured if I could get the view to load into a good/complete state then I can continue troubleshooting. 

Here's a quick video of the behavior I'm seeing. I'm not shocked that it's this "weird" in behavior but I'm out of ideas on how to get things to 'connect' more consistently. Here's [a gist of the current View and ViewModel](https://gist.github.com/houstonhaynes/f45b73a960d9b2e843494f84c12692a2). Thanks for giving it a once-over!

https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/70793ab4-93a8-4b18-a0c6-e9675268ae84

**Comment by JordanMarr** [comment]
What is it supposed to do? Or which properties are not updating properly?

**Reply by houstonhaynes** [reply]
There's a GIF in the first page I ref'd which shows the behavior. The "scrollbar" chart is a miniature snapshot of the full chart and it never "changes". What *does* change in that scrollbar is the "thumb" which is a area that's drawn over the scrollbar chart that matches the zoom area of the main chart. Grab that thumb with a mouse-down action and slide along the X axis and that shifts/updates the position of the the zoomed-in "MainChart" visual. So they interact with one another. It's a more rudimentary version of this working Highcharts(R) example on [this page](https://jkunst.com/highcharter/articles/highcharter.html). The LiveCharts example doesn't have "handles" on the thumb - yet 😉  - but I wanted to get the interactions from that sample under wraps before I started on extending it.

![image](https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/8174976/373056d5-5dcd-44a4-a77f-8888965da086)

**Reply by houstonhaynes** [reply]
The weird thing that's happening at the moment *seems* to be that the View isn't fully loading at first. I have to step "into" that view - which "to the eye" seems to load. I get *one* console event that the chart "updated" - but then - no other on-screen events/actions are picked up. Only *after* I terminate and re-start the view (by selecting another view in the nav menu and then coming back to that "ZoomView") do all of the mouse actions get "seen" in the MVU loop. At that *second* run of the view I start seeing all of the on-screen mouse events over the chart area - pointer up/down/moved and the "chart updated" events that occur when the X zoom level of the MainChart is changed by scrolling the mouse.  That's what you 're seeing when you watch the video clip.

I can keep working on the interaction between the two charts to make the behavior more cohesive between them. That's not the problem. The question I'm asking you is - have you've seen this "I have to go to the view twice before the MVU loop 'lights up'" behavior? And if so, is there anything that pops out to you in what I've done that might be a "land mine" I don' know about?

**Reply by Jamie-Rower** [reply]
I believe I encountered the same issue, and at least know why its happening.  The model is binding a list, but the desired change is to a property of an element in the list, and that is not being detected as a change to the list for the binding.  In my case I recreated the list instead of changing the elements, which worked briefly before other problems presented, so I'm fairly sure this is the cause.  I'm working on a solution

**Reply by JordanMarr** [reply]
It’s hard to tell from just looking at your sample bc it’s a lot of code. I did notice that you’re binding XAxes and YAxes that’s not actually a model property which seems sus.

**Reply by houstonhaynes** [reply]
> It’s hard to tell from just looking at your sample bc it’s a lot of code. I did notice that you’re binding XAxes and YAxes that’s not actually a model property which seems sus.

That's actually the kind of advice I was seeking. Thanks! We know it's a mess right now, but the point is to keep things "as local as possible" to reason over it in the View/code-behind and ViewModel.

I expect this will eventually become its own custom control, but I want to see it work first. The thing that bothers me is the example is so tight and terse while in our case slinging quite a bit of code, and that was my first sense of a "smell".

**Reply by Jamie-Rower** [reply]
So I made some progress on this.  First, the reactive commands were not dispatching properly because of a null value exception being thrown by initial value retrieval from the chart, specifically: let zoomlevel = xAxis.MaxLimit.Value.  I'm not sure why, but this only happened when running through the debugger.  Adding a null check before trying to access it not only cleared the exception, but in both debug and dotnet run the elmish messages started getting delivered properly.  Still not working, but the delivery system was all green.

The real irony is that in testing I removed zoomlevel from use to simplify, and the thumb is now resizing and moving properly as you zoom and pan the main chart.

The last leg is getting the top chart to update properly when dragging the thumb, and this, I think, is where my original thought comes back into play.

**Reply by Jamie-Rower** [reply]
There we go.  The axis in use for the main chart and the scroll thumb were not the same.  Rewired that and all appears to be working.

**Reply by Jamie-Rower** [reply]
https://github.com/JordanMarr/ReactiveElmish.Avalonia/assets/147878760/0e5e2673-4440-4a77-869a-6eb46b804c7a

**Reply by houstonhaynes** [reply]
Well post back a sample when we get the code in a state that's marginally less embarrassing (mostly my fault) 😁

**Reply by JordanMarr** [reply]
That would be awesome! 
I'd like to update the sample project to use the Fluent Avalonia menu.
